### PR TITLE
Abi jama

### PR DESCRIPTION
**Pull request by PerryBaran**

### CODE REVIEW

File: src/booleans.js [hunk 1]
```diff
@@ -1,62 +1,66 @@
...
+  return (a && !b) || (!a && b);
+}
 
 function truthiness(a) {
-  // your code here
-};
+  if (a) { 
+    return true; } 
+    else {
+    return false; }
```
Comment (PerryBaran) [inline src/booleans.js:25]: Instead of an `if` condition here, you could use a double bang `!` operator. The first bang operator would convert it to the opposite of its truthiness, the second would then flip that truthiness, leaving you with the truthiness of a, i.e.
```
return !!a
```

File: src/booleans.js [hunk 2]
```diff
@@ -1,62 +1,66 @@
...
-  // your code here
-};
+  return Math.sqrt(a) % 1 === 0;
+}
 
 function startsWith(char, string) {
-  // your code here
-};
+  return string[0] === char ? true : false;
+}
```
Comment (PerryBaran) [inline src/booleans.js:54]: Instead of using the tertiary operator `?`, and returning `true` or `false` depending on what the condition evaluates to, just return the condition. e.g.
```
return a % 2 !== 0
```

File: src/booleans.js [hunk 3]
```diff
@@ -1,62 +1,66 @@
...
 
 function containsVowels(string) {
-  // your code here
-};
+  return /[aeiouAEIOU]/.test(string);
```
Comment (PerryBaran) [inline src/booleans.js:57]: by adding a `i` after your regex string, you can make your test case insensitive. Meaning you only have to include the lower case vowels in your regex, i.e.
```
return /[aeiou]/i.test(string)
```

File: src/numbers.js [hunk 4]
```diff
@@ -1,46 +1,46 @@
...
+const quotient = (a, b) => {
+  return Math.trunc(a / b);
 }
 
-function remainder (a, b) {
-  // your code here
-}
+const remainder = (a, b) => {
+  return a % b;
+};
```
Comment (PerryBaran) [inline src/numbers.js:43]: All look good, but you're getting some linter/formatting warnings (red squiggly lines) due to missing semi colons. Although these won't affect how your code runs it is important to always follow the formatting set by your linter.

File: src/strings.js [hunk 5]
```diff
@@ -1,26 +1,26 @@
...
+function firstCharacter(string) {
+  return string.charAt(0);
+}
 
-function firstCharacters (string, n) {
-  // your code here
-};
+function firstCharacters(string, n) {
+  return string.slice(0, n);
+}
```
Comment (PerryBaran) [inline src/strings.js:23]: All perfect

File: src/objects.js [hunk 6]
```diff
@@ -1,43 +1,58 @@
...
 };
 
 const getProperty = (property, object) => {
-  // your code here
+  return object[property];
 };
 
 const hasProperty = (property, object) => {
-  // your code here
+  return !!object[property];
```
Comment (PerryBaran) [inline src/objects.js:14]: Although this code works for the tests provided, if the property on the object had a falsey value (e.g. `0`) then this would return `false` instead of `true`. A better solution would be to use the [`in`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/in) operator i.e.
```
return property in object
```

File: src/objects.js [hunk 7]
```diff
@@ -1,43 +1,58 @@
...
 const isOver65 = person => {
-  // your code here
+  return person.age>65;
 };
 
 const getAges = people => {
-  // your code here
-};
+  const array = [];
+  // eslint-disable-next-line no-plusplus
```
Comment (PerryBaran) [inline src/objects.js:23]: You should try avoid disabling eslint rules whenever possible. Instead, try to fix what's causing the issue. In this case, it wants you to use `i += 1` instead of `i++`.

File: src/objects.js [hunk 8]
```diff
@@ -1,43 +1,58 @@
...
+  for (let i = 0; i < people.length; i++) {
+    array.push(people[i].age);
+  }
+  return array;
+}
```
Comment (PerryBaran) [inline src/objects.js:28]: You could also tackle this problem using a `.map`, e.g.
```
return people.map(person => person.age)
```

File: src/objects.js [hunk 9]
```diff
@@ -1,43 +1,58 @@
...
 
 const averageAge = people => {
-  // your code here
-};
+  if (people.length === 0) {
+    return 0;
+  }
+  const totalAge = people.reduce((sum, person) => sum + person.age, 0);
+  const averageAge = totalAge / people.length;
+  return averageAge;
```
Comment (PerryBaran) [inline src/objects.js:44]: `averageAge` is also the name of the function. For readability sake you should think of a different name to call this variable. Alternatively, you could just return what you've assigned this this variable straight away. e.g.
```
return totalAge / people.length;
```

File: src/arrays.js [hunk 10]
```diff
@@ -1,61 +1,71 @@
...
 
 const addToArray = (element, array) => {
-  // your code here
+  array.push(element);
 };
 
 const addToArray2 = (element, array) => {
-  // your code here
+  const addArray = array.concat(element);
+  return addArray;
```
Comment (PerryBaran) [inline src/arrays.js:20]: Can just return the result instead of assigning to a variable first, i.e.
```
return array.concat(element)
```

File: src/arrays.js [hunk 11]
```diff
@@ -1,61 +1,71 @@
...
-  // your code here
-};
+  const vowels = ['a', 'e', 'i', 'o', 'u'];
+  return strings.filter(str => {
+    if (str.length === 0) {
+      return false;
+    }
+    const firstChar = str.charAt(0).toLowerCase();
+    return vowels.includes(firstChar);
+  });
```
Comment (PerryBaran) [inline src/arrays.js:54]: You could simplify this solution by using regex. e.g.
```
return strings.filter(string => string[0].match(/[aeiou]/i))
```

**Review by PerryBaran — COMMENTED** [review]
Overall really good. Lots of good solutions to the problems and I can see that you've thought about edge cases even when they are not included in the tests. I've left comments throughout on areas that could be improved.

There is a few linting/formatting problems throughout this project. If you have a red squiggly line in your code editor and the file name shows up in red make sure to fix these issues. If you hover the mouse of the red squiggly line a popup will appear with information about what's causing that issue with links to read more about it. At the bottom of this popup is usually some text that says `Quick Fix (Ctrl+.)` (Ctrl is a different symbol on mac). Clicking this button will bring up another popup with a few suggestions, the top one on this list will usually automatically fix your problems for you
